Challo/Challo/Common/Services/ImageService/ImageService.swift

```diff
@@ -0,0 +1,62 @@
+//
+//  ImageService.swift
+//  Challo
+//
+//  Created by Shao Yi on 10/4/21.
+//
+
+import SwiftUI
+import ImageKitIO
+import SDWebImageSwiftUI
+
+struct ImageService: ImageServiceProtocol {    
+        
+    private static func generateFullUrl(path: String, height: CGFloat, width: CGFloat) -> String {
+        ImageKit.shared.url(path: path, transformationPosition: .QUERY)
+            .height(height: Int(height))
+            .aspectRatio(width: Int(width), height: Int(height))
+            .create()
+    }
+    
+    static func loadImage(path: String, height: CGFloat = 400, width: CGFloat = 400,
+                          onSuccess: ((UIImage, Data?) -> Void)? = nil) -> Image {
+        let fullUrl = generateFullUrl(path: path, height: height, width: width)
+        
+        let image = WebImage(url: URL(string: fullUrl))
+            .onSuccess { image, data, _ in
+                onSuccess?(image, data)
+            }
+            .resizable()
+            .indicator(.activity)
+            .transition(.fade(duration: 0.5))
+//            .placeholder {
+//                Rectangle().foregroundColor(.gray)
+//            }
+        return Image(uiImage: image.asUIImage())
+    }
+        
+    static func uploadImage(image: Data, fileName: String, onProgress: ((Progress) -> Void)?,
+                            onSuccess: (() -> Void)?, onFailure: ((Error) -> Void)?) {
+        ImageKit.shared.uploader().upload(
+          file: image,
+          fileName: fileName,
+          useUniqueFilename: true,
+          folder: "/",
+          isPrivateFile: false,
+          customCoordinates: "",
+          responseFields: "",
+          progress: { progress in
+            onProgress?(progress)
+          },
+          completion: { result in
```

Comment:
Does the library give back a URL from which we can access the recently uploaded image as part of the `result`? If so then should we pass it back in the `onSuccess` callback? So can easily store it, e.g inside `UserState`. If not then how do we know the url to use ah?

Reply:
Technically cos Imagekit only needs the filename, so since clients of Imagekit provide the filename, we dont exactly have to return the full url. But yea i think we should anyways for conformity to the ImageService as an abstraction

Reply:
@shaoyi1997 Oh then are we enforcing the uniqueness of filenames on the client side somehow? cos I skimmed through the Imagekit [docs](https://docs.imagekit.io/api-reference/upload-file-api/server-side-file-upload) and there was this section

>useUniqueFileName: If set true, ImageKit.io will add a unique suffix to the filename parameter to get a unique filename.

Since `useUniqueFilename` is set to true in our case, if the filename already exists on the Imagekit server side, then the actual filename of it once it's saved on the server will be different. For e.g `abc.png` uploaded twice, then the second upload of `abc.png` maybe renamed by Imagekit to `abc123.png`?